### PR TITLE
Dai intel dmic additions

### PR DESCRIPTION
**Pull request by juimonen**
- add dmic dts definitions for ace platform
- add support in the code for tgl and apl platforms (add cavs25 and 15 dts dmic defs)
- misc dmic code fixes found in sof native driver compilation

### CODE REVIEW

**Review by dcpleung — DISMISSED** [review]

File: dts/xtensa/intel/intel_cavs15.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_cavs15.dtsi:293]: same note

File: dts/xtensa/intel/intel_cavs25.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_cavs25.dtsi:277]: ditto

File: drivers/dai/intel/dmic/dmic.c
Comment (andyross) [inline drivers/dai/intel/dmic/dmic.c:45]: I know this isn't really related to this PR and you're just moving this around, but this code is weirding me out.  The DMIC driver should be copying raw data from the device to an application layer.  If there is rescaling/transformation/resampling/whatever needed that really should be done at the app layer.  Zephyr doesn't (and shouldn't!) know anything about how PDM microphone signal processing works, that's what SOF is for.

**Review by andyross — COMMENTED** [review]
That address trick seems like an abuse of DTS (to my somewhat amateur eyes) and there should be cleaner ways to do what you want (including just using the DT_INST index, since you control the order in the dts file).  @galak @mbolivar might have better suggestions.

File: drivers/dai/intel/dmic/dmic.c
Comment (juimonen) [reply drivers/dai/intel/dmic/dmic.c:45]: I think the original reason to have it very close to hardware is that there's a "feature" in dmic hw, that makes a "pop" (unwanted spike in the waveform) in the very beginning. The dmic does some processing in hw ip, like decimation filtering etc. but there's no hw implementation for ramp or any other algorithm to get rid of the error.

So we could have a "ramp component" in sof pipeline, but what it would essentially do is to make a couple of millisecond ramp, then just pass through the 99,9% of the samples.

In this PR I was just moving this away from the header as it was giving me warning from dmic.nhlt.c

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:288]: yes correct. I just basically want to have 2 entries in dts to be able to instantiate 2 separate dmic dais. In practice it is the 1 and same hw block, but it can operate 2 streams at the same time (we can have for example 48kHz and 16kHz streams from same microphones at the same time).

Historically in sof these have been treated as separate dais, which I still think is the most clear thing to do. People start to get really confused when starting to talk about dmic internal pdm and fifo structures.

If someone has better suggestion, I will change this for sure.

Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:288]: @galak @mbolivar @ceolin any suggestions how to do this nicely the "correct way"? see my explanation above.

Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi]: @andyross tried to "clean" the dts by adding separating fifo field into the yaml.... not sure this is better/cleaner.

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: This seems reasonable to me (again: with "me" **not** being the best reference for DTS design patterns).  But AFAICT you're only ever using the first entry of this array in the driver code.  What's does the second value mean?  Also, is there a way to construct this from a simpler "index" instead of a raw register offset?  Maybe the hardware has microphone register blocks in an array with a stride of 256 bytes?

File: drivers/dai/intel/dmic/dmic.c
Comment (ceolin) [inline drivers/dai/intel/dmic/dmic.c:254]: where this variable comes from ?

File: drivers/dai/intel/dmic/dmic.c
Comment (ceolin) [inline drivers/dai/intel/dmic/dmic.c:272]: ditto

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (ceolin) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: Yep, what does the second value means ?

File: drivers/dai/intel/dmic/dmic.c
Comment (juimonen) [reply drivers/dai/intel/dmic/dmic.c:254]: copy paste error from sof code... removing

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: Still nitpicky: you dropped the second item, but now it's an array of one element.  You can have integer properties ("type: int"), which seems like what you really want.

Also I still think this would be clearer if it stored an index (0 or 1) instead of a register offset.

(Not a -1, just a note)

File: drivers/dai/intel/dmic/dmic.c
Comment (mbolivar-nordic) [inline drivers/dai/intel/dmic/dmic.c:868]: I think I agree with @andyross; this should be an int if you only expect one value inside. Unless there's something I'm missing? The commit logs don't have much information and there is no property description documenting what this is supposed to contain, so I'm going to -1 until we can clear this up.

If you switch to int, this just becomes `DT_INST_PROP(n, fifo)`.

**Review by mbolivar-nordic — DISMISSED** [review]

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: np, good to get feedback. I just noticed that zephyr is silently merging the 2 nodes with same register address.. ??? So I don't get 2 devices, only 1. So I probably need to come up with some other scheme here for the nodes.

Comment (mbolivar-nordic) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: > So I probably need to come up with some other scheme here for the nodes.

Assuming that they really are two different peripherals that just happen to have the same register map addresses, you can just give them different names. Examples:

https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/dts/arm/nordic/nrf52840.dtsi#L87
https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/dts/arm/nordic/nrf52840.dtsi#L104

This will cause a warning which you can disable with https://github.com/zephyrproject-rtos/zephyr/blob/6f9e8f459d261f36dc4cf7e50376f4a4931e39ea/boards/arm/nrf52840dk_nrf52840/pre_dt_board.cmake#L7.

Comment (mbolivar-nordic) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:283]: ... although you did just find a bug in dtlib. I assumed `dtc` would have the same behavior, but it actually treats this as an error by default:

```
$ cat foo.dts
/dts-v1/;

/ {
	blub: foo@1234 {
		reg = <1234 4>;
		bar = <1>;
	};

	blab: foo@1234 {
		reg = <1234 4>;
		baz = <2>;
	};
};

$ dtc -Odts foo.dts
foo.dts:8.17-10.4: ERROR (duplicate_node_names): /foo@1234: Duplicate node name
ERROR: Input tree has errors, aborting (use -f to force output)
```

We are compatible with dtc by contract, so this should have been an error. I'll file a bug, thank you for noticing this.

Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi]: @mbolivar-nordic I tried to do this as you suggested... hopefully kind of what you were asking

File: drivers/dai/intel/dmic/dmic.c
Comment (juimonen) [reply drivers/dai/intel/dmic/dmic.c]: ok, should be like that now

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
Comment (juimonen) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi]: changed now to int value

File: drivers/dai/intel/dmic/dmic.c
Comment (juimonen) [reply drivers/dai/intel/dmic/dmic.c]: ok, fixed now

**Review by dcpleung — DISMISSED** [review]

File: drivers/dai/intel/dmic/dmic.c
Comment (mbolivar-nordic) [reply drivers/dai/intel/dmic/dmic.c]: >  The commit logs don't have much information and there is no property description documenting what this is supposed to contain, so I'm going to -1 until we can clear this up.

This was a hint to improve the commit log and add a property description, just FYI. I'm going to dismiss my review now because it's moved to `int`, but I think these should be improved.

**Comment by juimonen** [comment]
@nashif can't really is this PR causing the unit test failures...? seems unrelated.

**Comment by juimonen** [comment]
@abonislawski @marcinszkudlinski @dcpleung @kv2019i  sorry to spam, but rebased now and CI clean. So need re-approvers if found ok :) thanks

**Review by dcpleung — DISMISSED** [review]

**Review by kv2019i — DISMISSED** [review]
Thanks @juimonen , looks good!

File: drivers/dai/intel/dmic/dmic.c
Comment (jxstelter) [inline drivers/dai/intel/dmic/dmic.c:253]: I think it would be good to have shim registers definitions and basic access functions separated from main DMIC driver code and moved to platform specific headers. This patch now adds #ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V15, but shim registers definition for SOC_SERIES_INTEL_ACE will change from one version to another.  In result soon we would have here several #ifdefs  for different generations.

**Comment by dcpleung** [comment]
Could you rebase to get rid of the merge commit?

**Comment by juimonen** [comment]
> Could you rebase to get rid of the merge commit?

yeah sorry, I think I accidentally pushed the update branch button here in github...

File: drivers/dai/intel/dmic/dmic.c
Comment (kv2019i) [reply drivers/dai/intel/dmic/dmic.c:253]: @juimonen @jxstelter Is this something to do now, or later in separate PR. Given we only have two cases now, I'd prefer to do this in separate PR.

**Review by kv2019i — APPROVED** [review]

File: drivers/dai/intel/dmic/dmic.c
Comment (jxstelter) [reply drivers/dai/intel/dmic/dmic.c:253]: OK.  Separate PR will be good.

**Review by dcpleung — APPROVED** [review]

**Review by abonislawski — APPROVED** [review]

**Pull request merged**
